imap_processing/ultra/l1b/de.py

```diff
@@ -13,46 +30,172 @@ def calculate_de(de_dataset: xr.Dataset, name: str) -> xr.Dataset:
     Parameters
     ----------
     de_dataset : xarray.Dataset
-        Dataset containing direct event data.
+        L1a dataset containing direct event data.
     name : str
-        Name of the dataset.
+        Name of the l1a dataset.
 
     Returns
     -------
     dataset : xarray.Dataset
-        Dataset containing the data.
+        L1b de dataset.
     """
     de_dict = {}
+    sensor = parse_filename_like(name)["sensor"][0:2]
 
-    # Placeholder for calculations
-    epoch = de_dataset.coords["epoch"].values
+    # Instantiate arrays
+    yf = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    xb = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    yb = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    xc = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    d = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float64)
+    r = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    tof = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    etof = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    ctof = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    energy = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.float32)
+    species_bin = np.full(len(de_dataset["epoch"]), np.nan, dtype=np.uint8)
```

Comment:
I'm curious what value this fills with. `np.nan` should only be valid for floating point types AFAIK.

Reply:
I was curious too - looks like np fills with 0s:
```
np.full((3,2), np.nan, dtype=np.uint8)
>>> array([[0, 0],
       [0, 0],
       [0, 0]], dtype=uint8)
```

Reply:
Good catch. I'll put a TODO for asking the Ultra team if zeros are acceptable. As of now they are still trying to figure out what they want in the species bin. They have said they might even want string values so this is all preliminary.